### PR TITLE
[ADD][12.0] crm_lead_currency

### PR DESCRIPTION
**Pull request by TDu**
This module allows to select a specific currency for an opportunity.
If the selected currency is different to the company currency,
an amount in the customer currency can be set and the expected revenue
of the opportunity will automatically be computed in the company currency.

### CODE REVIEW

File: crm_opportunity_currency/README.rst [hunk 1]
```diff
@@ -0,0 +1,77 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (simahawk) [inline crm_opportunity_currency/README.rst]: pls use new readme structure for new modules

File: crm_opportunity_currency/README.rst [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+========================
+Crm Opportunity Currency
+========================
+
+This module allows to select a specific currency for an opportunity.
+If the selected currency is different to the company currency, an amount in the customer
```
Comment (simahawk) [inline crm_opportunity_currency/README.rst]: *different from

File: crm_opportunity_currency/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    'name': 'CRM Opportunity Currency',
+    'summary': """
+        On opportunities add the amount in the company currency.""",
+    'version': '12.0.1.0.0',
+    'license': 'AGPL-3',
+    'author': 'Camptocamp SA,Odoo Community Association (OCA)',
+    'website': 'www.camptocamp.com',
```
Comment (simahawk) [inline crm_opportunity_currency/__manifest__.py]: https://github.com/OCA/crm

File: crm_opportunity_currency/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    'depends': [
+        'crm',
+    ],
+    'data': [
+        'views/crm_lead_opportunity_currency_views.xml',
+    ],
+    'demo': [
```
Comment (simahawk) [inline crm_opportunity_currency/__manifest__.py]: remove empty key

File: crm_opportunity_currency/models/crm_lead.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    @api.multi
+    def get_revenue_in_company_currency(self):
+        """Compute the planned revenue in the company currency.
+
+        If the customer currency is different than the company currency,
+        the planned revenue is computed in the company currency.
+        """
+        self.ensure_one()
+        if self.company_currency == self.customer_currency_id:
```
Comment (simahawk) [inline crm_opportunity_currency/models/crm_lead.py]: `if self.is_same_currency`?

File: crm_opportunity_currency/models/crm_lead.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        currency_field='customer_currency_id',
+        store=True,
+    )
+    is_same_currency = fields.Boolean(
+        string='Same currency',
+        compute='_compute_is_same_currency',
+    )
+
+    @api.onchange('customer_currency_id', 'amount_customer_currency')
+    def compute_expected_amount(self):
```
Comment (simahawk) [inline crm_opportunity_currency/models/crm_lead.py]: `_onchange_currency`

File: crm_opportunity_currency/models/crm_lead.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            )
+
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        if (
+            'customer_currency_id' in vals
+            or 'amount_customer_currency' in vals
+        ):
+            # Due to the readonly in the view the compute does not work
```
Comment (simahawk) [inline crm_opportunity_currency/models/crm_lead.py]: I'd remove the readonly... I don't see any bad side effect... if they edit it, up to them to fix it, no?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: crm_opportunity_currency/README.rst (hunk 2 above)
Comment (TDu) [reply crm_opportunity_currency/README.rst]: I think both are valid english

File: crm_opportunity_currency/__manifest__.py (hunk 3 above)
Comment (TDu) [reply crm_opportunity_currency/__manifest__.py]: Changed

File: crm_opportunity_currency/__manifest__.py (hunk 4 above)
Comment (TDu) [reply crm_opportunity_currency/__manifest__.py]: Done

File: crm_opportunity_currency/models/crm_lead.py (hunk 5 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: Yes

File: crm_opportunity_currency/models/crm_lead.py (hunk 6 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: Changed

File: crm_opportunity_currency/README.rst (hunk 1 above)
Comment (TDu) [reply crm_opportunity_currency/README.rst]: Changed

File: crm_opportunity_currency/models/crm_lead.py (hunk 7 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: I would think that if it is not set to readonly the currency conversion should be handled in both direction.

Comment (i-vyshnevska) [reply crm_opportunity_currency/models/crm_lead.py]: why don't relay planned_revenue on compute method

**Review by i-vyshnevska — COMMENTED** [review]
would be good to send choosen currency from lead to SO, but it would call dependency from 'sale'

File: crm_opportunity_currency/models/crm_lead.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from odoo import api, fields, models
+
+
+class CrmLead(models.Model):
+    _inherit = 'crm.lead'
+
+    customer_currency_id = fields.Many2one(
+        string='Customer currency',
```
Comment (luistorresm) [inline crm_opportunity_currency/models/crm_lead.py]: In v12 when the string is the same that the field name, this parameter is not necessary.

And if it has `_id` or `_ids`, that is not added in the string.

Could you remove this parameter, please?

File: crm_opportunity_currency/models/crm_lead.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        store=True,
+    )
+    is_same_currency = fields.Boolean(
+        string='Same currency',
+        compute='_compute_is_same_currency',
+    )
+
+    @api.onchange('customer_currency_id', 'amount_customer_currency')
+    def _onchange_currency(self):
+        for lead in self:
```
Comment (luistorresm) [inline crm_opportunity_currency/models/crm_lead.py]: In the onchange method is not necessary to make for, this is not multi.

[Here](https://www.odoo.com/documentation/10.0/reference/orm.html#odoo.api.onchange) the Odoo doc

File: crm_opportunity_currency/models/crm_lead.py (hunk 8 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: But it doesn't hurt to have it explicitly stated ?

File: crm_opportunity_currency/models/crm_lead.py (hunk 9 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: Thanks changed.

File: crm_opportunity_currency/models/crm_lead.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+    customer_currency_id = fields.Many2one(
+        string='Customer currency',
+        comodel_name='res.currency',
+        default=lambda self: self.env.user.company_id.currency_id,
+    )
+    amount_customer_currency = fields.Monetary(
+        string='Customer amount',
+        currency_field='customer_currency_id',
+        store=True,
```
Comment (grindtildeath) [inline crm_opportunity_currency/models/crm_lead.py]: I don't think this is necessary if the field is not computed.

File: crm_opportunity_currency/models/crm_lead.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        the planned revenue is computed in the company currency.
+        """
+        self.ensure_one()
+        if self.is_same_currency:
+            return self.planned_revenue
+        return self.customer_currency_id._convert(
+            self.amount_customer_currency or 0,
+            self.company_currency,
+            self.env.user.company_id,
+            fields.Datetime.now(),
```
Comment (grindtildeath) [inline crm_opportunity_currency/models/crm_lead.py]: Would be great to somehow be able to change the date used here, but a simple mention in the roadmap section of the readme is fine at the moment.

File: crm_opportunity_currency/models/crm_lead.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        if (
+            'customer_currency_id' in vals
+            or 'amount_customer_currency' in vals
+        ):
+            # Due to the readonly in the view the compute does not work
+            for lead in self:
+                lead.write({
```
Comment (grindtildeath) [inline crm_opportunity_currency/models/crm_lead.py]: Calling a write inside a loop inside a write is usually a bad design.
I'd add `customer_currency_id=None` and `amount_customer_currency=None` to the params of `get_revenue_in_company_currency`. Then use `or` in the method to get it right, eg:

```
customer_currency = self.env['res.currency'].browse(customer_currency_id) or self.customer_currency_id
return customer_currency._convert(
    amount_customer_currency or self.amount_customer_currency or 0,
    (...)
```
Then move the loop before the call to super and update `vals` for each record, e.g.:
``` 
@api.multi
def write(self, vals):
    if ('customer_currency_id' in vals or 'amount_customer_currency' in vals):
        for lead in self:
            planned_revenue = lead.get_revenue_in_company_currency(
                customer_currency_id=vals.get('customer_currency_id')),
                amount_customer_currency=vals.get('amount_customer_currency')
            )
            vals['planned_revenue'] = planned_revenue
            super(CrmLead, lead).write(vals)
        return True
    else:
        return super().write(vals)
```

File: crm_opportunity_currency/views/crm_lead_opportunity_currency_views.xml [hunk 13]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0"?>
+
+<odoo>
+
+  <record id="crm_case_form_view_oppor" model="ir.ui.view">
```
Comment (grindtildeath) [inline crm_opportunity_currency/views/crm_lead_opportunity_currency_views.xml]: Somehow these views should be restricted to group `base.group_multi_currency`

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Small changes

File: crm_opportunity_currency/models/crm_lead.py (hunk 10 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: Correct left over from previous code. Removed now

File: crm_opportunity_currency/models/crm_lead.py (hunk 11 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: Readme updated.

File: crm_opportunity_currency/views/crm_lead_opportunity_currency_views.xml (hunk 13 above)
Comment (TDu) [reply crm_opportunity_currency/views/crm_lead_opportunity_currency_views.xml]: I am not sure it seems to me, that group is mainly used in accounting module to restrict some access. Here I think if the module is installed we want every user to have access to this feature.

File: crm_opportunity_currency/models/crm_lead.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+from odoo import api, fields, models
+
+
+class CrmLead(models.Model):
+    _inherit = 'crm.lead'
+
+    customer_currency_id = fields.Many2one(
+        string='Customer currency',
+        comodel_name='res.currency',
+        default=lambda self: self.env.user.company_id.currency_id,
```
Comment (sebalix) [inline crm_opportunity_currency/models/crm_lead.py]: Maybe add `ondelete='restrict`, what do you think? (the default value is `set null`)

File: crm_opportunity_currency/models/crm_lead.py (hunk 8 above)
Comment (gurneyalex) [reply crm_opportunity_currency/models/crm_lead.py]: it does not hurt at all

File: crm_opportunity_currency/models/crm_lead.py (hunk 14 above)
Comment (gurneyalex) [reply crm_opportunity_currency/models/crm_lead.py]: @sebalix yes (but in this specific case, chances are really high that some other required currency_id field on another model will prevent deleting the currency record, which is not something you do very often :smile_cat: )

File: crm_opportunity_currency/models/crm_lead.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            return self.planned_revenue
+        return self.customer_currency_id._convert(
+            self.amount_customer_currency or 0,
+            self.company_currency,
+            self.env.user.company_id,
+            fields.Datetime.now(),
+        )
+
+    @api.multi
+    def _compute_is_same_currency(self):
```
Comment (gurneyalex) [inline crm_opportunity_currency/models/crm_lead.py]: needs changing: missing api.depends decorator -> the field is not recomputed when the currency is changed -> the field amount_customer_currency is not visible until you save the record

File: crm_opportunity_currency/models/crm_lead.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+        for lead in self:
+            lead.is_same_currency = (
+                lead.customer_currency_id == lead.company_currency
```
Comment (gurneyalex) [inline crm_opportunity_currency/models/crm_lead.py]: The test must be updated to return True if customer_currency_id is not set (in this case we assume it means we are using the company currency)

**Review by gurneyalex — CHANGES_REQUESTED** [review]

**Comment by pedrobaeza** [comment]
Following model naming, this should be `crm_lead_currency`

**Comment by gurneyalex** [comment]
renaming requested by Pedro is in https://github.com/TDu/crm/pull/1 (with a small improvement in the manifest file)

File: crm_opportunity_currency/models/crm_lead.py (hunk 15 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: This has been done

File: crm_opportunity_currency/models/crm_lead.py (hunk 12 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: The change on the write has been removed following changes in the view to handle the problem with the readonly field.

File: crm_opportunity_currency/models/crm_lead.py (hunk 16 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: @gurneyalex I have made a slight variation see the last commit

**Comment by TDu** [comment]
The module name has been changed I squashed the commit a little as well.

File: crm_opportunity_currency/models/crm_lead.py (hunk 16 above)
Comment (TDu) [reply crm_opportunity_currency/models/crm_lead.py]: It allows the module to works the same in creation or update mode.

**Review by gurneyalex — APPROVED** [review]

**Review by i-vyshnevska — APPROVED** [review]

**Review by claudiagn — APPROVED** [review]
LGTM!

**Pull request merged**